### PR TITLE
Fix output/input of Arrow dictionary column

### LINKED ISSUE #54927: output_format_arrow_low_cardinality_as_dictionary produces invalid output
**Describe what's wrong**

When using the `Arrow` format in conjunction with `LowCardinality(String)`, invalid output is produced when the number of distinct values is 128 or more.

This issue is a continuation of #36832.

**Does it reproduce on recent release?**

Yes: v23.8.2.7-lts

**How to reproduce**

Using the method in [#37043](../pull/37043#issuecomment-1139570143), we can reproduce it by inserting precisely 128 distinct values:

```sql
$ clickhouse-client --version
ClickHouse client version 23.8.2.7 (official build).

:) select version()

SELECT version()

Query id: a61aac5f-b076-44ea-b00b-0e405088b810

┌─version()─┐
│ 23.8.2.7  │
└───────────┘

1 row in set. Elapsed: 0.002 sec.

:) insert into function file(data.arrow) select toLowCardinality(toString(number)) as x from numbers(128) settings output_format_arrow_low_cardinality_as_dictionary=1

INSERT INTO FUNCTION file(data.arrow)
SETTINGS output_format_arrow_low_cardinality_as_dictionary = 1
SELECT toLowCardinality(toString(number)) AS x
FROM numbers(128)
SETTINGS output_format_arrow_low_cardinality_as_dictionary = 1

Query id: a8af0b58-f2d3-496b-8940-413f2312dcfb

Ok.

0 rows in set. Elapsed: 0.005 sec.

:) select uniq(x) from file(data.arrow)

SELECT uniq(x)
FROM file(data.arrow)

Query id: 28e86982-a82e-4703-9961-7f313df25bbe

┌─uniq(x)─┐
│      65 │
└─────────┘

1 row in set. Elapsed: 0.007 sec.
```

**Expected behavior**

`uniq(x)` should be 128 not 65.

**Additional context**

I suspect the issue is related to the arrow schema inference, specifically this [function](../blob/c8619ee6e4acaf7718a790d15e03261df4d4d3a7/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp#LL799C58-L799C58) where the indices type is inferred. The function incorrectly maps unsigned types to signed types of the same width (e.g. `arrow::int8()` cannot represent the same range as `TypeIndex::UInt8`).

### LINKED ISSUE #55490: ArrowStream with different dictionaries in every batch
script used to generate example.arrowstream ([example.zip](https://github.com/ClickHouse/ClickHouse/files/12867311/example.zip))

```py
import numpy as np
import pandas as pd
import pyarrow as pa

batch1 = pa.record_batch([pa.array(["foo"], type=pa.dictionary(pa.int32(), pa.utf8()))], "x")
batch2 = pa.record_batch([pa.array(["foo", "bar"], type=pa.dictionary(pa.int32(), pa.utf8()))], "x")

with pa.ipc.new_stream("example.arrowstream", schema=batch1.schema) as stream:
  stream.write(batch1)
  stream.write(batch2)

print(pa.ipc.open_stream("example.arrowstream").read_pandas())
```

As you can see the file consists of 2 batches, every of those batches have DIFFERENT dictionaries.

But ClickHouse loads the dict only once, and that leads to exception:
 
```
SELECT * FROM file('/tmp/example.arrowstream')

┌─x───┐
│ foo │
└─────┘

Received exception:
Code: 117. DB::Exception: Index 1 in Dictionary column is out of bounds, dictionary size is 1: While executing ArrowBlockInputFormat: While executing File. (INCORRECT_DATA)
```

ClickHouse doesn’t handle the dictionary changing in the stream (which is valid for Arrow IPC streams, though not for IPC files).
https://github.com/ClickHouse/ClickHouse/blob/5ad055df98a69d7a487e96895cd9e6b16e7ba740/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp#L900

Worse still, if the new dictionary has similar or fewer elements, ClickHouse can silently give the wrong answer
[example2.zip](https://github.com/ClickHouse/ClickHouse/files/12867497/example2.zip)

<details><summary>script used to generate example.arrowstream</summary>

```py
import numpy as np
import pandas as pd
import pyarrow as pa

batch1 = pa.record_batch([pa.array(["foo", "XXX"], type=pa.dictionary(pa.int32(), pa.utf8()))], "x")
batch2 = pa.record_batch([pa.array(["foo", "bar"], type=pa.dictionary(pa.int32(), pa.utf8()))], "x")

with pa.ipc.new_stream("example2.arrowstream", schema=batch1.schema) as stream:
  stream.write(batch1)
  stream.write(batch2)

print(pa.ipc.open_stream("example2.arrowstream").read_pandas())
```

</details>

the correct output is:

```
     x
0  foo
1  XXX
2  foo
3  bar
```

but ClickHouse prints:

```
SELECT *
FROM file('/tmp/example2.arrowstream')

Query id: 59fffb50-b5a8-47e0-9c39-7f279390fe3e

┌─x───┐
│ foo │
│ XXX │
└─────┘
┌─x───┐
│ foo │
│ XXX │
└─────┘

4 rows in set. Elapsed: 0.005 sec. 
```

Additionally - ArrowStream should support appending 

```
INSERT INTO FUNCTION file('test.arrowstream', 'ArrowStream') SELECT *
FROM values('a LowCardinality(String)', ('foo', 'bar'))


INSERT INTO FUNCTION file('test.arrowstream', 'ArrowStream') SELECT *
FROM values('a LowCardinality(String)', ('foo', 'baz'))


Received exception:
Code: 641. DB::Exception: Cannot append data in format ArrowStream to file, because this format doesn't support appends. You can allow to create a new file on each insert by enabling setting engine_file_allow_create_multiple_files. (CANNOT_APPEND_TO_FILE)
```

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix incorrect indexes type selection for Arrow dictionary during output. Fix reading Arrow dictionaries in ArrowStream format
Closes https://github.com/ClickHouse/ClickHouse/issues/55490
Closes https://github.com/ClickHouse/ClickHouse/issues/54927

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit c080ba9d7eb8074913ba028d21eb97661c95e0dc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55989/c080ba9d7eb8074913ba028d21eb97661c95e0dc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6631834918">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55989/c080ba9d7eb8074913ba028d21eb97661c95e0dc/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp [hunk 1]
```diff
@@ -356,6 +357,18 @@ namespace DB
         }
     }
 
+    static void checkIfIndexesTypeIsExceeded(const std::shared_ptr<arrow::DataType> & arrow_type, size_t dict_size)
+    {
+        const auto & dict_indexes_arrow_type = assert_cast<const arrow::DictionaryType *>(arrow_type.get())->index_type();
+        /// We use UInt32 or UInt64 type for indexes. It makes sense to check overflow only for UInt32.
+        const auto * indexes_uint32_type = typeid_cast<const arrow::UInt32Type *>(dict_indexes_arrow_type.get());
```
Comment (vdimir) [inline src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:364]: Alternatively we can compare `dict_indexes_arrow_type->bit_width()` and `std::bit_width(dict_size)`, but I don't think it's worth changing. The only reason is that if we change `getArrowTypeForLowCardinalityIndexes` one more time (e.g. to use uint16)  we will need not to forget this place.

**Review by vdimir — APPROVED** [review]

**Comment by den-crane** [comment]
@Avogar 

Now it fails

```
rm -f output.arrows && ./clickhouse local -nq "SET output_format_arrow_low_cardinality_as_dictionary = 1; SELECT CAST('foo', 'LowCardinality(String)') AS str INTO OUTFILE 'output.arrows' FORMAT ArrowStream" && python3 -c 'import pyarrow as pa; print(pa.ipc.open_stream("output.arrows").read_all().to_pandas())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "pyarrow/array.pxi", line 884, in pyarrow.lib._PandasConvertible.to_pandas
  File "pyarrow/table.pxi", line 4192, in pyarrow.lib.Table._to_pandas
  File "/usr/local/lib/python3.11/site-packages/pyarrow/pandas_compat.py", line 774, in table_to_blockmanager
    blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyarrow/pandas_compat.py", line 1122, in _table_to_blocks
    result = pa.lib.table_to_blocks(options, block_table, categories,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow/table.pxi", line 3115, in pyarrow.lib.table_to_blocks
  File "pyarrow/error.pxi", line 91, in pyarrow.lib.check_status
pyarrow.lib.ArrowTypeError: Converting unsigned dictionary indices to pandas not yet supported, index type: uint32
```

**Comment by Avogar** [comment]
Well, it's a problem with pandas that it doesn't support unsigned integers as indices. But I can introduce a setting for it, there will be no problem

**Pull request merged**